src/main/scala/nl/vroste/zio/kinesis/client/DynamicConsumer.scala

```diff
@@ -28,237 +18,47 @@ import software.amazon.kinesis.exceptions.ShutdownException
-            kinesisClient,
-            dynamoDbClient,
-            cloudWatchClient,
-            workerIdentifier,
-            () => new ZioShardProcessor(queues)
-          )
-          leaseTableName.fold(configsBuilder)(configsBuilder.tableName)
-        }
-        env           <- ZIO.environment[R].toManaged_
-
-        scheduler <- Task(
-                       new Scheduler(
-                         configsBuilder.checkpointConfig(),
-                         configsBuilder.coordinatorConfig(),
-                         configsBuilder.leaseManagementConfig().initialPositionInStream(initialPosition),
-                         configsBuilder.lifecycleConfig(),
-                         configsBuilder.metricsConfig(),
-                         configsBuilder.processorConfig(),
-                         configsBuilder
-                           .retrievalConfig()
-                           .initialPositionInStreamExtended(initialPosition)
-                           .retrievalSpecificConfig(retrievalConfig(kinesisClient))
-                       )
-                     ).toManaged_
-        doShutdown = // UIO(println("Starting graceful shutdown")) *>
-                     ZIO.fromFutureJava(scheduler.startGracefulShutdown()).unit.orDie <*
-                       queues.shutdown
-        _         <- zio.blocking
-               .blocking(ZIO(scheduler.run()))
-               .fork
-               .flatMap(_.join)
-               .onInterrupt(doShutdown)
-               .forkManaged
-        _         <- (requestShutdown *> doShutdown).forkManaged
-      } yield ZStream
-        .fromQueue(queues.shards)
-        .collectWhileSuccess
-        .map {
-          case (shardId, shardQueue, checkpointer) =>
-            val stream = ZStream
-              .fromQueue(shardQueue.q)
-              .ensuringFirst(shardQueue.shutdownQueue)
-              .collectWhileSuccess
-              .mapChunksM(_.mapM(toRecord(shardId, _)))
-              .provide(env)
-              .ensuringFirst(checkpointer.checkpoint.catchSome { case _: ShutdownException => UIO.unit }.orDie)
-
-            (shardId, stream, checkpointer)
-        }
-
-    ZStream.unwrapManaged(schedulerM)
+    def shardedStream[R, T](
+      streamName: String,
+      applicationName: String,
+      deserializer: Deserializer[R, T], // TODO make Deserializer a ZLayer too
+      requestShutdown: UIO[Unit] = UIO.never,
+      initialPosition: InitialPositionInStreamExtended =
+        InitialPositionInStreamExtended.newInitialPosition(InitialPositionInStream.TRIM_HORIZON),
+      isEnhancedFanOut: Boolean = true,
+      leaseTableName: Option[String] = None,
+      workerIdentifier: String = UUID.randomUUID().toString,
+      maxShardBufferSize: Int = 1024    // Prefer powers of 2
+    ): ZStream[
+      Blocking with R,
+      Throwable,
+      (String, ZStream[Blocking, Throwable, Record[T]], Checkpointer)
+    ]                                   // TODO - what about Blocking?
```

Comment:
What about it?

Reply:
I think when I wrote the comment I had a vague notion of pushing this dependency further up but think about it now I do not think this will be possible. I will delete the comment

Reply:
done